File: app/models/Tag.java
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package models;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.ManyToMany;
import jakarta.persistence.Table;
import jakarta.persistence.UniqueConstraint;
import java.util.List;
import models.base.OwnedModel;
import models.questions.Question;
import org.apache.commons.lang3.builder.EqualsBuilder;
import org.apache.commons.lang3.builder.HashCodeBuilder;

@Entity
@Table(uniqueConstraints = @UniqueConstraint(columnNames = { "name", "creator_id" }))
public class Tag extends OwnedModel {

    @Column(nullable = false, length = 32)
    private String name;

    @ManyToMany(mappedBy = "tags", cascade = CascadeType.ALL)
    private List<Question> questions;

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public List<Question> getQuestions() {
        return questions;
    }

    public void setQuestions(List<Question> questions) {
        this.questions = questions;
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) return true;
        if (!(other instanceof Tag otherTag)) {
            return false;
        }
        return new EqualsBuilder().append(name, otherTag.name).build();
    }

    @Override
    public int hashCode() {
        return new HashCodeBuilder().append(name).build();
    }
}

File: app/models/questions/Tag.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package models.questions

import jakarta.persistence.*
import models.base.OwnedModel

import scala.compiletime.uninitialized

@Entity
@Table(uniqueConstraints = Array(new UniqueConstraint(columnNames = Array("name", "creator_id"))))
class Tag extends OwnedModel:
  @ManyToMany(mappedBy = "tags", cascade = Array(CascadeType.ALL))
  var questions: java.util.List[Question] = uninitialized

  var name: String = uninitialized

  override def equals(o: Any): Boolean = o match
    case t: Tag => this.name == t.name
    case _      => false

  override def hashCode: Int = if name != null then name.hashCode else 0
